Il2CppInspector.Common/Outputs/IDAPythonScript.cs

```diff
@@ -72,8 +72,10 @@ public class IDAPythonScript
         private void writeMethods() {
             writeSectionHeader("Method definitions");
             foreach (var type in model.Types) {
-                writeMethods(type.Name, type.DeclaredConstructors);
-                writeMethods(type.Name, type.DeclaredMethods);
+                if (type != null) {
```

Comment:
Would it not be preferable to avoid adding nulls to the collection in the first place?

Reply:
The way we did this previously was to add a "broken" `TypeInfo` that did not represent a real type (i.e. it was one that was constructed with `new` but had no fields added). I am not convinced that's much better - a TypeRef with `-1` for a definition index means a missing type, so `null` seems the most appropriate.

Previously, I did not have any test files that contained such an empty TypeRef, but it turns out the CSharp3 test case does include an empty TypeRef, so it should be possible in principle to figure out why such references get included in the first place.

Reply:
It is very strange that there are occasional files with these non-existent references.

I have to say though that I really do not like having nulls in non-associative collections; for readable code I think it is reasonable for an end user to be able to assume that the expected behaviour when fetching a collection is that every item in the collection will contain an object of the collection's type (non-null).

So rather than adding a null check here (which would have to be replicated anywhere anyone uses the collection), can we suppress adding/returning nulls in the first place? I understand why we have it in TypesByReferenceIndex where you can use a reference index as the array key, but Types is basically just a flat list and I don't feel like random null values in the middle of the list make sense.

Reply:
I think I have a hypothesis for where these null references come from - they appear to be created when the generic recursion depth overflows. `MaximumRuntimeGenericDepth` is set to 8, which is pretty low and is actually hit in practice easily by Linq queries: the CSharp3 testcase generates 46 of these (and I checked - every such null reference corresponds to a class with 8 levels of nested generics). For example, one of the classes for which the type reference is null is:

`System.Collections.Generic.IComparer`1<System.Linq.Parallel.Pair`2<System.Linq.Parallel.Pair`2<System.Linq.Parallel.Pair`2<System.Linq.Parallel.Pair`2<System.Linq.Parallel.Pair`2<System.Linq.Parallel.Pair`2<System.Object,System.Object>,System.Object>,System.Object>,System.Object>,System.Object>,System.Object>>`

I am not sure why it's "ok" for the type reference to be null. `GenericMetadata.cpp` contains the rather ominous warning

```
        // As we do for code generation, let's cut this off at an arbitrary level. If something tries to execute code at this
        // level, a crash will happen. We'll assume that this code won't actually be executed though.
```

but since I haven't delved too far into the code generation side of things, I'm not sure if code generation is OK with a nesting depth this deep.

Reply:
`MaximumRuntimeGenericDepth` is an il2cpp.exe compilation parameter I assume, or some hard-coded thing in it? Very interesting indeed. Feel free to add a comment about that in the code somewhere if you like for future reference (optional).

Reply:
Regardless - agreed that `Types` shouldn't contain null pointers - so I'll fix the PR to do that instead.

Reply:
`MaximumRuntimeGenericDepth` is in GenericMetadata.h (so it's a hardcoded thing in the Il2Cpp runtime itself), but I also just noticed that `il2cpp.exe` has a similar argument `--maximum-recursive-generic-depth=` which defaults to 7.